Improve Memory Allocations

wolveix opened the pull request:
Hey! Firstly, great project, we rely on it for some of our core data processing :)

We noticed an issue with our project where memory wasn't being returned from the heap. It ended up being due to holding strings returned by `ParseMany`, more specifically the `parseObjects` function. We were able to fix this on our end by explicitly nulling data after use, but I thought I'd take some time to dive into your library to see if I could address the issue.

This PR:
- Fixes the heap issue (by slightly reworking the `parseObjects` function
- Reduces the memory allocations from the tests from 35M objects down to 19.5M objects (`newAttribute` was the main culprit)
- Adds a test to cover the `object.GetFirst` function
- Enhances a few internal errors with a little more context
- Updates exposed function comments to fix the [Go docs guidelines](https://tip.golang.org/doc/comment)

Original library memory profile (from tests):
<img width="676" alt="Screenshot 2025-03-19 at 19 19 25" src="https://github.com/user-attachments/assets/68aff8ea-8a5a-4b95-b6b5-c0fe945de79d" />

Revised library memory profile (from tests):
<img width="665" alt="image" src="https://github.com/user-attachments/assets/11fe3674-4c07-4d2a-ae9c-76899e0ebcb0" />

From our testing, this looks good and doesn't cause any regressions. But please do test it, and let me know if anything should be changed!

wolveix committed:
Fix memory alloc issue
Add minor code documentation
Provide more verbose errors
Add test for `Object.GetFirst`

wolveix committed:
Additional attribute optimizations
Accept streaming objects from a reader

wolveix commented:
I just added another commit which further optimized the attribute function, but more importantly added a `ParseManyFromReader` function. I think this should actually replace the current `ParseMany` function as it's much more efficient; however, I didn't want to change your package's API without first getting your thoughts :) But in our project test, it dropped our allocs from 60GB down to 43GB.

frederic-arr commented:
Hi, thanks for the PR! Everything looks good on my end.

> I just added another commit which further optimized the attribute function, but more importantly added a ParseManyFromReader function. I think this should actually replace the current ParseMany function as it's much more efficient; however, I didn't want to change your package's API without first getting your thoughts :) But in our project test, it dropped our allocs from 60GB down to 43GB.

I don't mind introducing that method but `ParseMany` needs to remain for backwards compatibility reasons. I sense a code duplication by keeping the two so replacing the body of `ParseMany` to a call to `ParseManyFromReader` may be necessary. If we follow this approach `parseObjects` should be removed too as it would be otherwise unused.

---

If this sounds good to you, and if you have time, making the required changes would be much appreciated but I can take them on otherwise.

wolveix committed:
Refactor to use bytes + 95% test coverage

wolveix commented:
Completely agree regarding keeping the API as-is!

Okay, so this ended up being a bigger task than initially expected. When switching the internal functions to use the new byte parsing, it ended up raising the total allocs back to around 35M unexpectedly. I figured what the heck, let's try rewriting the library to use bytes wherever possible, to avoid unnecessary type conversations and lighter parsing internally. While doing so, I added additional tests to cover the majority of the library (I believe total coverage is now 95%).

Tests do naturally fluctuate from one run to the next, and while I was able to get the new version to be lighter than the previous commit (~17M allocs), I suspect they're approximately the same. Interestingly when comparing the memory profiles, the new version is heavier with the `newAttribute` function, but the rest of the functions essentially do nothing from an alloc-perspective:

<img width="878" alt="image" src="https://github.com/user-attachments/assets/d5f3a051-4be8-48e6-8027-150df0da61bd" />

Versus the previous commit:

<img width="878" alt="image" src="https://github.com/user-attachments/assets/d0e9de76-d58b-4ba3-9a77-3ef12a4f84e5" />

---

I added a benchmark function to give a more tangible point of comparison between the various revisions:

Original repo:
```
BenchmarkParse-12    	  833626	      1313 ns/op
```

Previous commit:
```
BenchmarkParse-12    	  932636	      1158 ns/op
```

New commit:
```
BenchmarkParseBytes-12    	 1000000	      1048 ns/op
```

---

Edit: it seems I got so caught up doing this that I forgot all about the streaming aspect 🤦🏻‍♂️ I'm reworking this now

wolveix commented:
Back to the drawing board :) I'm not too sure why the stream reader causes allocs to increase so drastically, it should be the other way round

wolveix committed:
Refactor to use `io.Reader`

wolveix commented:
This took a lot of back and forth to figure out the optimal implementation, but I think I've found it. The benchmark shows that parsing is faster than before:

```
BenchmarkParseFromReader-12    	 1288480	       914.8 ns/op
```

Memory allocs are slightly higher within the library (~19M), but allow the caller to drastically reduce their allocs; in our project, being able to use `io.Reader` drops our initial sync from 60GB allocs down to just 27GB (60GB was from the previous commit).

As tests fluctuate from one run to another, I was able to see the memory profile as low as 16.9M (but it averages around 19M):
![image](https://github.com/user-attachments/assets/8b950966-3803-4ae3-b8f7-048ce99d6a66)

---

Hopefully this looks good, but let me know if you want anything revised!

wolveix committed:
Optimize `newAttribute`

wolveix commented:
Figured I'd revisit the `newAttribute` function as that's where the majority of the heavy allocs occur. I was able to further optimize it:

```
BenchmarkParseFromReader-12    	 1369954	       849.3 ns/op
```

Memory profiles are now averaging 18M again :)

wolveix committed:
Micro optimizations

wolveix commented:
Okay, I think I've squeezed every bit of performance I can at this point.

```
BenchmarkParseFromReader-12    	 1547264	       760.8 ns/op
```

Memory allocs remain roughly the same as before, but I also made a few micro optimizations:
```
Attribute.String() optimization:
28ns -> 15 ns/op

Object.String() optimization:
104ns -> 75 ns/op
```

Thanks for making such a great library, and hopefully there are no issues with these changes!

wolveix committed:
Formatting

wolveix committed:
Formatting

frederic-arr commented:
Everything looks good to me, thanks for taking the time to make the changes and thanks for your contribution!

(I'll merge the changes in a couple of hours)

wolveix committed:
Add back missing comments

wolveix commented:
Seems I accidentally removed a few comments I didn't mean to, so I've just added those back :) Thanks so much!

frederic-arr commented:
Changes have been merged, thanks again for your contribution!
